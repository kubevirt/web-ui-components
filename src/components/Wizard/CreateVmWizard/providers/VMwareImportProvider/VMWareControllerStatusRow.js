import React from 'react';
import PropTypes from 'prop-types';

import { Alert, Col, FormGroup, Spinner } from 'patternfly-react';

import { Link } from 'react-router-dom';

import {
  getV2vVMwareDeploymentStatus,
  V2V_WMWARE_DEPLOYMENT_STATUS_FAILED,
  V2V_WMWARE_DEPLOYMENT_STATUS_POD_FAILED,
  V2V_WMWARE_DEPLOYMENT_STATUS_PROGRESSING,
  V2V_WMWARE_DEPLOYMENT_STATUS_ROLLOUT_COMPLETE,
  V2V_WMWARE_DEPLOYMENT_STATUS_UNKNOWN,
} from '../../../../../utils/status/v2vVMwareDeployment';

import VMwareStatusField from './VMwareStatusField';
import { FormRow } from '../../../../Form/FormRow';
import { getName } from '../../../../../selectors';

import { getSubPagePath } from '../../../../../utils';
import { DeploymentModel, PodModel } from '../../../../../models';

const DeploymentLink = ({ deployment }) => (
  <Link to={getSubPagePath(deployment, DeploymentModel, 'events')}>{getName(deployment)}</Link>
);

DeploymentLink.defaultProps = {
  deployment: null,
};

DeploymentLink.propTypes = {
  deployment: PropTypes.object,
};

const NoDeployment = () => (
  <VMwareStatusField>
    Starting VMware controller... <Spinner loading size="sm" />
  </VMwareStatusField>
);

const DeploymentProgressing = ({ id, deployment }) => (
  <VMwareStatusField id={id}>
    Deploying VMware controller (<DeploymentLink deployment={deployment} />
    )... <Spinner loading size="sm" />
  </VMwareStatusField>
);

DeploymentProgressing.defaultProps = {
  deployment: null,
  id: null,
};

DeploymentProgressing.propTypes = {
  deployment: PropTypes.object,
  id: PropTypes.string,
};

const DeploymentFailed = ({ deployment, pod, message }) => {
  let podMessage;
  if (pod) {
    podMessage = (
      <React.Fragment>
        {' '}
        Please inspect a failing pod <Link to={getSubPagePath(pod, PodModel, 'events')}>{getName(pod)}</Link>
      </React.Fragment>
    );
  }
  return (
    <VMwareStatusField>
      <Alert type="warning">
        Deployment of VMware controller <DeploymentLink deployment={deployment} /> failed: {message}.{podMessage}
      </Alert>
    </VMwareStatusField>
  );
};

DeploymentFailed.defaultProps = {
  ...DeploymentProgressing.defaultProps,
  pod: null,
  message: null,
};

DeploymentFailed.propTypes = {
  ...DeploymentProgressing.propTypes,
  pod: PropTypes.object,
  message: PropTypes.string,
};

const vmwareStatusComponentResolver = {
  [V2V_WMWARE_DEPLOYMENT_STATUS_PROGRESSING]: DeploymentProgressing,
  [V2V_WMWARE_DEPLOYMENT_STATUS_POD_FAILED]: DeploymentFailed,
  [V2V_WMWARE_DEPLOYMENT_STATUS_FAILED]: DeploymentFailed,
};

const VMWareControllerStatusRow = props => {
  const { id, hasErrors, deployment, deploymentPods, controlSize, labelSize } = props;

  const status = getV2vVMwareDeploymentStatus(deployment, deploymentPods);
  if (
    status.status === V2V_WMWARE_DEPLOYMENT_STATUS_ROLLOUT_COMPLETE ||
    (status.status === V2V_WMWARE_DEPLOYMENT_STATUS_UNKNOWN && hasErrors) // deployment failed
  ) {
    return null;
  }

  const StatusComponent = vmwareStatusComponentResolver[status.status] || NoDeployment;

  return (
    <FormGroup>
      <Col sm={labelSize} />
      <Col sm={controlSize}>
        <StatusComponent id={id} {...status} />
      </Col>
    </FormGroup>
  );
};

VMWareControllerStatusRow.defaultProps = {
  id: null,
  hasErrors: false,
  deployment: null,
  deploymentPods: null,
  controlSize: FormRow.defaultProps.controlSize,
  labelSize: FormRow.defaultProps.labelSize,
};

VMWareControllerStatusRow.propTypes = {
  id: PropTypes.string,
  hasErrors: PropTypes.bool,
  deployment: PropTypes.object,
  deploymentPods: PropTypes.array,
  controlSize: FormRow.propTypes.controlSize,
  labelSize: FormRow.propTypes.labelSize,
};

export default VMWareControllerStatusRow;
