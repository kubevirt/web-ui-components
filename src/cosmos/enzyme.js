import createContext from 'react-cosmos-test/enzyme';

import paths from '../../config/paths';

export default function createTestContext(args) {
  args.proxies = args.proxies || []; // workaround of react-cosmos-test bug
  return createContext({
    ...args,
    cosmosConfigPath: `${paths.config}/cosmos.config.js`,
  });
}
