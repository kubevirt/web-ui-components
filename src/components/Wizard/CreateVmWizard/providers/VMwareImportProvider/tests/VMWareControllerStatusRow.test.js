import React from 'react';
import { render } from 'enzyme';

import v2vVMwareDeploymentFixtures from '../../../../../../utils/status/v2vVMwareDeployment/fixtures/v2vVMwareDeploymentStatus.fixture';
import VMWareControllerStatusRow from '../VMWareControllerStatusRow';
import { createContext } from '../../../../../../tests/router';

describe('<VMWareControllerStatusRow deployment deploymentPods />', () => {
  v2vVMwareDeploymentFixtures.forEach((fixture, idx) => {
    const { expected, deployment, deploymentPods } = fixture;
    it(`renders correctly ${idx} (${expected})`, () => {
      expect(
        render(<VMWareControllerStatusRow deployment={deployment} deploymentPods={deploymentPods} />, createContext())
      ).toMatchSnapshot();
    });
  });
});
