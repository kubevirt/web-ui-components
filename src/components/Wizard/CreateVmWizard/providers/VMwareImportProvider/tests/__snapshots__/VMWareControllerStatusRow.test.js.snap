// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<VMWareControllerStatusRow deployment deploymentPods /> renders correctly 0 (V2V_WMWARE_DEPLOYMENT_STATUS_UNKNOWN) 1`] = `
<div
  class="form-group"
>
  <div
    class="col-sm-3"
  />
  <div
    class="col-sm-5"
  >
    <div
      class="kubevirt-create-vm-wizard__import-vmware-connection-status"
    >
      Starting VMware controller... 
      <div
        class="spinner spinner-sm"
      />
    </div>
  </div>
</div>
`;

exports[`<VMWareControllerStatusRow deployment deploymentPods /> renders correctly 1 (V2V_WMWARE_DEPLOYMENT_STATUS_PROGRESSING) 1`] = `
<div
  class="form-group"
>
  <div
    class="col-sm-3"
  />
  <div
    class="col-sm-5"
  >
    <div
      class="kubevirt-create-vm-wizard__import-vmware-connection-status"
    >
      Deploying VMware controller (
      <a
        href="/k8s/ns/kube-public/deployments/v2v-vmware/events"
      >
        v2v-vmware
      </a>
      )... 
      <div
        class="spinner spinner-sm"
      />
    </div>
  </div>
</div>
`;

exports[`<VMWareControllerStatusRow deployment deploymentPods /> renders correctly 2 (V2V_WMWARE_DEPLOYMENT_STATUS_FAILED) 1`] = `
<div
  class="form-group"
>
  <div
    class="col-sm-3"
  />
  <div
    class="col-sm-5"
  >
    <div
      class="kubevirt-create-vm-wizard__import-vmware-connection-status"
    >
      <div
        class="alert alert-warning"
      >
        <span
          aria-hidden="true"
          class="pficon pficon-warning-triangle-o"
        />
        Deployment of VMware controller 
        <a
          href="/k8s/ns/kube-public/deployments/v2v-vmware/events"
        >
          v2v-vmware
        </a>
         failed: ProgressDeadlineExceeded.
      </div>
    </div>
  </div>
</div>
`;

exports[`<VMWareControllerStatusRow deployment deploymentPods /> renders correctly 3 (V2V_WMWARE_DEPLOYMENT_STATUS_POD_FAILED) 1`] = `
<div
  class="form-group"
>
  <div
    class="col-sm-3"
  />
  <div
    class="col-sm-5"
  >
    <div
      class="kubevirt-create-vm-wizard__import-vmware-connection-status"
    >
      <div
        class="alert alert-warning"
      >
        <span
          aria-hidden="true"
          class="pficon pficon-warning-triangle-o"
        />
        Deployment of VMware controller 
        <a
          href="/k8s/ns/kube-public/deployments/v2v-vmware/events"
        >
          v2v-vmware
        </a>
         failed: Back-off pulling image "quay.io/kubevirt/kubevirt-vmware:latests". Please inspect a failing pod 
        <a
          href="/k8s/ns/kube-public/pods/v2v-vmware-676769c6c8-rr4q9/events"
        >
          v2v-vmware-676769c6c8-rr4q9
        </a>
      </div>
    </div>
  </div>
</div>
`;

exports[`<VMWareControllerStatusRow deployment deploymentPods /> renders correctly 4 (V2V_WMWARE_DEPLOYMENT_STATUS_ROLLOUT_COMPLETE) 1`] = `null`;
